subsys/bluetooth/host/mesh/settings.c

```diff
@@ -499,102 +473,97 @@ static int hb_pub_set(int argc, char **argv, char *val)
 	return 0;
 }
 
-static int cfg_set(int argc, char **argv, char *val)
+static int cfg_set(int argc, char **argv, void *val_ctx)
 {
 	struct bt_mesh_cfg_srv *cfg = bt_mesh_cfg_get();
-	int len, err;
-
-	BT_DBG("val %s", val ? val : "(null)");
+	int err;
 
 	if (!cfg) {
 		return -ENOENT;
 	}
 
-	if (!val) {
+	if (settings_val_get_len_cb(val_ctx) == 0) {
+		BT_DBG("val (null)");
 		stored_cfg.valid = false;
 		BT_DBG("Cleared configuration state");
 		return 0;
 	}
 
-	len = sizeof(stored_cfg.cfg);
-	err = settings_bytes_from_str(val, &stored_cfg.cfg, &len);
+
+	err = mesh_x_set(val_ctx, &stored_cfg.cfg, sizeof(stored_cfg.cfg));
 	if (err) {
-		BT_ERR("Failed to decode value %s (err %d)", val, err);
+		BT_ERR("Failed to set \'cfg\'");
 		return err;
 	}
 
-	if (len != sizeof(stored_cfg.cfg)) {
-		BT_ERR("Unexpected value length (%d != %zu)", len,
-		       sizeof(stored_cfg.cfg));
-		return -EINVAL;
-	}
-
 	stored_cfg.valid = true;
 	BT_DBG("Restored configuration state");
 
 	return 0;
 }
 
-static int mod_set_bind(struct bt_mesh_model *mod, char *val)
+static int mod_set_bind(struct bt_mesh_model *mod, void *val_ctx)
 {
-	int len, err, i;
+	int len, i;
 
 	/* Start with empty array regardless of cleared or set value */
 	for (i = 0; i < ARRAY_SIZE(mod->keys); i++) {
 		mod->keys[i] = BT_MESH_KEY_UNUSED;
 	}
 
-	if (!val) {
+	if (settings_val_get_len_cb(val_ctx) == 0) {
```

Comment:
If this is supposed to signify a deletion, isn't it simpler to just use NULL?

Reply:
I was thinking about idea you signal here as well. Unfortunately have not time to implement this return NULL feature. I rather will provide this as other path (which will be compatible backward). I also found API for get  value length usefull.

Reply:
How do you differentiate between zero-length values and deleted entries? I thought that's why the original API was using NULL for this?

Reply:
It was all the same - NULL meant 0 length and vice versa.

Reply:
So there has never been a way to indicate that an entry was deleted?

Reply:
I'm a bit confused now, since AFAIK the settings subsystem has a way of doing garbage collection/compression by eliminating deleted entries when reaching the end of the partition. In such a case you need to be able to tell apart valid but zero-length entries from entries that were deleted (i.e. can be discarded).

Reply:
Issue clarified in slack: no way to differentiate between deleted and zero-length entries. Should be documented, but not a new issue, so could be fixed in separate PR.